+sw_tests/+unit_tests/unittest_spinw_spinwave.m

```diff
@@ -0,0 +1,471 @@
+                                                4.5  4.5  9];
+            expected_sw.gtensor = true;
+            testCase.verify_spinwave(expected_sw, sw_g, 'rel_tol', 1e-15);
+        end
+        function test_hermit(testCase)
+            % Create copy to avoid changing obj for other tests
+            swobj_h = copy(testCase.swobj);
+            % Tests that the 'hermit' option to switch to a non-hermitian calculation works
+            % First make the model non-Hermitian by adding a large axial SIA perpendicular to the spins
+            swobj_h.addmatrix('label', 'K', 'value', diag([-1 0 0]));
+            swobj_h.addaniso('K');
+            hkl = {[0 0 0] [0 1 0] [1 0 0] 50};
+            % Check that calling it with 'hermit' on gives an error
+            testCase.assertError(@() swobj_h.spinwave(hkl, 'hermit', true), ?MException);
+            % Now check that there are imaginary eigenvalues/energies in the output with 'hermit' off
+            spec = swobj_h.spinwave(hkl, 'hermit', false);
+            testCase.assertGreaterThan(sum(abs(imag(spec.omega(:)))), 0);
+        end
+        function test_sw_qh5_tol(testCase)
+            tol = 5e-4;
+            qpts = testCase.qh5;
+            swobj_tol = copy(testCase.swobj);
+            % Generate magstr that deviates slightly from commensurate
+            swobj_tol.genmagstr('mode', 'helical', 'k', [tol 0 0], ...
+                                'n', [0 0 1], 'S', [0; 1; 0]);
+            % Check that without tol it is incommensurate - causes error
+            testCase.verifyError(...
+                @() swobj_tol.spinwave(qpts), ...
+                'spinw:spinwave:NonPosDefHamiltonian');
+            % Check that with tol is approximated to commensurate
+            sw_out = swobj_tol.spinwave(qpts, 'tol', tol);
+            expected_sw = testCase.get_expected_sw_qh5;
+            expected_sw.obj = swobj_tol;
+            expected_sw.param.tol = tol;
+            testCase.verify_spinwave(expected_sw, sw_out);
+        end
+        function test_sw_qh5_omega_tol(testCase)
+            qpts = testCase.qh5;
+            % Check that with no added omega_tol Hamiltonian isn't positive
+            % definite - causes error
+            testCase.verifyError(...
+                @() testCase.swobj.spinwave(qpts, 'omega_tol', 0), ...
+                'spinw:spinwave:NonPosDefHamiltonian');
+            % Check with omega_tol the omega is changed appropriately
+            omega_tol = 1;
+            sw_out = testCase.swobj.spinwave(qpts, 'omega_tol', omega_tol);
+            expected_sw = testCase.get_expected_sw_qh5;
+            expected_sw.omega(:, 1) = [omega_tol -omega_tol];
+            expected_sw.omega(:, end) = [-omega_tol omega_tol];
+            expected_sw.param.omega_tol = omega_tol;
+            testCase.verify_spinwave(expected_sw, sw_out);
+        end
+        function test_biquadratic_with_incomm_causes_error(testCase)
```

Comment:
Yes, that's what I'm testing for right? That biquadratic and k != 0 causes an error?

Reply:
The docs would imply that it should also throw an error for a non-zero commensurate k as well right?

Reply:
Oh, I thought incommensurate == non-zero k... I said my magnetism was bad!

Reply:
I think perhaps it's OK just to rename this test (and any others) to say non-zero k. Maybe we should have a separate test for an incommensurate modulation though, if that is indeed supported?

Reply:
Umm... actually it's not necessary that `k=[0,0,0]` exactly, any integer value of `k` is fine, this would also result in a commensurate calculation. The problem is that the "incommensurate" (or rather the rotating coordinate frame) calculation is not implemented for biquadratic interactions. The "k=0" / "k!=0" in the comment is just a short-hand for commensurate / incommensurate (sorry for the sloppiness).

Below is a bit of a digression on the maths... feel free to ignore it!

So the biquadratic interaction is defined as:

![\mathcal{H}_{\mathrm{bq}} = \sum_{ij} K_{ij} \left( \mathbf{S}_i \cdot \mathbf{S}_j \right)^2](https://latex.codecogs.com/png.image?\dpi{110}\mathcal{H}_{\mathrm{bq}}&space;=&space;\sum_{ij}&space;K_{ij}&space;\left(&space;\mathbf{S}_i&space;\cdot&space;\mathbf{S}_j&space;\right)^2)

Where unlike in the bilinear case (e.g. the usual case we deal with) the `K` coupling constant is a scalar (not a 3x3 tensor) because the term in brackets is a scalar - you can express the dot product as `Si*Sj*cos(theta)` so the potential due to the biquadratic interaction is dependent on the _angle_ `theta` between spins `i` and `j`. This is where [the dot products in the code](https://github.com/SpinW/spinw/blob/master/swfiles/%40spinw/spinwave.m#L573-L577) come from I think.

Now, in the rotating coordinate frame, we have to take into account the small rotations between the unit cells in order to calculate the angle between the spins `i` and `j`. And this is currently not implemented. I think it might be a bit tricky, hence Sandor has not done it, but it should be possible in principle. (For the bilinear case because the couplings are _not_ dependent on the relative angles between the spins we can just [apply a rotation matrix](https://github.com/SpinW/spinw/blob/master/swfiles/%40spinw/spinwave.m#L519-L525) to them - this is eq (21) in the Toth/Lake paper.)